torch/csrc/jit/passes/utils/memory_dag.cpp

```diff
@@ -116,59 +119,51 @@ Element* MemoryDAG::makeFreshValue(const Value* v) {
   return rawPtr;
 }
 
-std::unordered_set<const Element*> Element::getMemoryLocations() const {
+const MemoryLocations& Element::getMemoryLocations() const {
   if (!cachedMemoryLocations_.empty()) {
     return cachedMemoryLocations_;
   }
 
   // Do a BFS in the `points-to` direction, collecting all memory locations
-  std::unordered_set<const Element*> ret;
-  this->bfs(
-      [&](const Element* el) {
-        if (el->pointsTo.empty()) {
-          ret.insert(el);
-        }
-      },
-      BfsDirection::POINTS_TO);
-
+  MemoryLocations ret;
+  this->bfs(BfsDirection::POINTS_TO, ret);
   cachedMemoryLocations_ = ret;
-  return ret;
+  return cachedMemoryLocations_;
 }
 
 // Do a breadth-first search over the graph, starting at `this` and
 // traversing in the direction `dir`.`fn` will be run on each element.
-template <typename Fn>
-bool Element::bfs(Fn fn, BfsDirection dir) const {
-  std::queue<const Element*> queue;
-  std::unordered_set<const Element*> seen;
-
-  queue.push(this);
+void Element::bfs(BfsDirection dir, MemoryLocations& res) const {
+  std::queue<unsigned> queue;
+  MemoryLocations seen;
```

Comment:
I suspect that this one wants to remain an `unordered_set`.  We're not using it for anything where a bit vector would be faster.

Reply:
I think what I've gotten from my benchmarking is that the primary speedup doesn't come from using bitset operations, it comes from the reduction in memory allocation (one bit instead of an entire node).

Reply:
Seems plausible.  I'd encourage you to test this specific part of the change to verify.

Reply:
"As a general statement, testing/setting bits in a SparseBitVector is O(distance away from last set bit)."